File: trunk/src/networking/TokenRing.java
package networking;

import java.io.IOException;
import java.util.ArrayList;
import java.util.Collections;
import java.util.LinkedList;
import java.util.List;
import java.util.Vector;

import org.apache.xmlrpc.XmlRpcException;

import networking.Client.RemoteNode;

public class TokenRing implements Runnable {
		
	private static TokenRing instance = null;
	
	private List<RemoteNode> network;
	private int indexInRing = -1;
	
	private String ip;
	private String ipCordinator;
	
	private Token token = null;
	
	public static TokenRing getInstance()
	{
		return instance;
	}
	
	public TokenRing(List<RemoteNode> network, String ip)
	{
		this.network = new LinkedList<RemoteNode>();
		this.network.addAll(network);
		this.ip      = ip;
		
		Collections.sort(network);
		
		for ( int index = 0; index < network.size(); index++ )
		{
			if ( network.get(index).ip.compareTo(this.ip) == 0 )
			{
				this.indexInRing = index;
				break;
			}
		}
		
		fetchCordinator();
		this.instance = this;
	}
	private void fetchCordinator() {
		ipCordinator = network.get(0).ip;
		if ( ipCordinator.compareTo(this.ip) == 0 )
		{
			// this node won the election according to IP metric:
			token = new Token( this.ip, this.ip );
			System.out.println("Won the election, created the ring!");
		}
	}
	

	
	@Override
	public void run() {
		while (true)
		{
			if ( isHolderOfToken() && isCalcDone() )
			{
				forwardToken();
			}
		}
	}
	private boolean isCalcDone() {
		// TODO Auto-generated method stub
		return true;
	}
	private boolean isHolderOfToken() {
		if ( token != null && token.currentHolder.compareTo(this.ip) == 0 )
		{
			return true;
		}
		return false;
	}
	
	private void forwardToken()
	{
		// fetch next peer:
		int nextPeer = ( indexInRing + 1 ) % network.size();
		System.out.println("Forwarding token to IP "+network.get(nextPeer).ip);

		Vector<String> params = new Vector<String>();

		params.add(token.ipCreator);
		params.add(network.get(nextPeer).ip);
		
		token = null;
		
		// send request to node
		try {
			network.get(nextPeer).rpc.execute("TokenRingAux.tokenReceived", params );
		} catch (XmlRpcException | IOException e) {
			// TODO Auto-generated catch block
			e.printStackTrace();
		}
	}
	
	public void receivedToken(Token token)
	{
		System.out.println("Received Token!");
		this.token = token;
	}

}

File: trunk/CSharp/TokenRing.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using CookComputing.XmlRpc;

namespace Networking
{
    class TokenRing : SyncAlgorithm
    {
        private int indexInRing = -1;
        private string ipCordinator;
        private volatile Token token = null;

        bool _keepRunning = true;


	
        public TokenRing(IList<RemoteNode> network, string ip) : base(network, ip)
		{
			// pay attension to the difference between network and this.network
       
            token = null;
           
            Console.WriteLine(Client.getInstance().getTable());
			this.network.Sort();

            this.indexInRing = Client.getInstance().SearchTableIndex(this.ip,this.network);

            /*
            for (int index = 0; index < this.network.Count; index++)
            {
                //Console.WriteLine(this.network[index].getIP());
				if (this.network[index].getIP().CompareTo(this.ip) == 0)
				{
					this.indexInRing = index;
					break;
				}
			}*/

            Console.WriteLine("My index in network is: " + this.indexInRing);
			fetchCordinator();
			TokenRing.instance = this;



		}

        public void RequestStop()
        {
            _keepRunning = false;
        }

		private void fetchCordinator()
		{
			ipCordinator = this.network[0].getIP();
			if (ipCordinator.CompareTo(this.ip) == 0)
			{
				// this node won the election according to IP metric:
				token = new Token(this.ip, this.ip);
				Console.WriteLine("Won the election, created the ring!");
			}
		}



		public void run()
        {
            Console.WriteLine("\nThread  TokenRing is running\n");
            while (!getcanStart());
            
            while (_keepRunning)
			{

				if ((hasRing() && CalcDone) && !isPending )
				{
					forwardToken();
				}
			}
            Console.WriteLine("\nThread  TokenRing is terminated\n");
		}



 

		public virtual bool hasRing()
		{
			lock (this)
			{
				if (token != null)
                {
                    if (token.currentHolder.CompareTo(this.ip) == 0)
					      return true;
				}
				return false;
			}
		}




		private void forwardToken()
		{
			// fetch next peer index:
		 int nextPeerIndex = (indexInRing + 1) %this.network.Count;
         String nextIP=this.network[nextPeerIndex].getIP();

         token.currentHolder = nextIP;	
	     Console.WriteLine("Forwarding token to IP "+nextIP );

	     // send request to node

         NetworkClientInterface executer = XmlRpcProxyGen.Create<NetworkClientInterface>();
         executer.AttachLogger(new XmlRpcDebugLogger());

         executer.Url = network[nextPeerIndex].getURL();
		 executer.tokenReceived(token.ipCreator,nextIP);

		}



		public virtual void receivedToken(Token token)
		{
			lock (this)
			{
			Console.WriteLine("Received Token!");
				this.token = token;
			}
		}
		public override bool canAccess()
		{
			lock (this)
			{
				// TODO Auto-generated method stub
				return hasRing();
			}
		}

	}

}



	
	
	

	
